library/ssl_srv.c

```diff
@@ -100,7 +100,7 @@ static int ssl_parse_servername_ext( mbedtls_ssl_context *ssl,
         return( MBEDTLS_ERR_SSL_BAD_HS_CLIENT_HELLO );
     }
     servername_list_size = ( ( buf[0] << 8 ) | ( buf[1] ) );
-    if( servername_list_size + 2 != len )
+    if( (size_t) servername_list_size + 2 != len )
```

Comment:
Changing the type of `servername_list_size` doesn't really make sense if you have to then do a cast. You can't change the type of `len` since it's in the function spec so it's not worth changing this instance.

Reply:
The key to this change is the previous line - two bytes make a 16-bit unsigned integer only; no additional space is required while `size_t` could be 32 or 64 bits.
As for type casts - you pointed out this issue previously, and its was necessary to suppress the compiler warning.
This cast should not generate additional code, because internally compiler would do about the same - expand 16 bits to integer size, and to `size_t` to compare.

Reply:
It wasn't me who pointed that out previously but anyway, in that earlier PR the cast was from a larger type to a smaller one whereas here you're casting to a larger type, which presumably will undo any code size reduction by declaring the smaller type? Have you actually demonstrated this improves code size in some cases?

Reply:
> here you're casting to a larger type

The existing code has a cast to a larger type too, an implicit one:
`servername_list_size = (size_t)( ( buf[0] << 8 ) | ( buf[1] ) );`

Reply:
Right. I'm not so much concerned about whether the cast is implicit or explicit. I'm more concerned the change won't improve the code, whether it's readability or code size or something else.

Reply:
My guess is that readability gets a tiny improvement,  because for `uint16_t` nobody would attempt to check if the value is below 2^16-1 (as was fixed in the one recently merged PR).
As was said previously: we have 16 bits of data and use exactly 16-bit variable to store it.

Code size is less predictable, unfortunately. Quite often usage of natural sizes such as `int` generates shorter compiled code even if in manual code it could easily be the opposite.

Reply:
Compilation of `ssl_srv.c` with `make no_test` using gcc in Ubuntu 20.04 64.bit created 55.5 K `ssl_srv.o` and 55.4 K for the file with the suggested changes.

Reply:
But that includes the other changes, which I do accept are an improvement. What is the difference when only considering these changes that involve casting back to a larger type?

Reply:
(I haven't followed all the discussions in this PR) Code size at this micro level is very dependent on the target architecture as well as the compiler version and the surrounding code. It can happen that using a 16-bit variable to store a 16-bit value is a win if the variable has its own register, but a loss if the variable is spilled to memory, which is highly sensitive to how many registers are available at this point in the code and how the compiler uses them.

x86_64 code size is not that great an indicator because it's a 64-bit platform and code size is usually not an issue there. When I'm concerned about code size, I do a Cortex-M0+ build: that's the second-smallest Arm platform, and the smallest (Cortex-M0).
```
make CC=arm-none-eabi-gcc AR=arm-none-eabi-ar LD=arm-none-eabi-ld
size library/*.o
```

About this specific line, a way to avoid a cast is
```
    if( servername_list_size + 2u != len )
```
I'd prefer to avoid a cast because a cast tells the reader “watch out, something fancy is happening here”, but nothing remarkable is happening on this line: there's no risk of overflow or truncation. The reason for the warning in https://github.com/ARMmbed/mbedtls/pull/3238/commits/919af9be696e980a3d8b66afa68d2023936ed555 is that an unsigned 16-bit type got widened to `int` (because it fit) which then got wrapped around to a `size_t`, and the compiler analyzed the wrapped-around part independently from the widening part and is telling us that negative values could suffer, but thanks to the first part we know that negative values are impossible. All of which the reader shouldn't have to care about, because the only reason negative numbers started popping up was the use of a type (`uint16_t`) that's smaller than `int` and C's promotion rules kicking in.

Reply:
@gilles-peskine-arm
Thanks for the idea; I made these changes and eliminated type casts.
On the subject of code size, it might be interesting if you could find time and check what it gets for these Cortex processor. Just for this single type change at line 90 to `uint16_t`.
For x86 and x64 code optimizer in MSVC is not doing a great job, and code with `size_t` was shorter.

Reply:
Compiler:
```
arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 7-2018-q3-update) 7.3.1 20180622 (release) [ARM/embedded-7-branch revision 261907]
```
Flags: `-Os -mthumb -mcpu=cortex-m0plus`
Library configuration: `scripts/config.py baremetal`

Code size of `ssl_srv.o` (`text` column of `size library/ssl_srv.o`):
* Originally (5cac33722878e51b764d571e40a627264f922b43): 20102
* Current version of this PR (a7025a215423413a252b0d5399979a65f4b4b54b): 20102
* Current version, but with `size_t` instead of `uint16_t` at line 90: 20106

I haven't analyzed the generated assembly.

Reply:
Thanks for the test.
Your numbers together with my data confirm that the difference is small and compiler-specific. Hence, the code size should be of a minor importance to prefer `size_t` to `uint16_t` or vice versa.

Reply:
By the way, there was no `All checks have passed` messages for a long time. Good to see it fixed.